### PR TITLE
How to build a Q&A web application using Python and Anvil

### LINKED ISSUE #6263: [Machine Learning] How to build a Q&A web application using Python and Anvil
#### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. 

How to build a Q&A web application using Python and Anvil

#### Proposed article introduction
*We're looking for the first 2-3 paragraphs of your article that appropriately summarize what your article will be about.*
We will use this writing snippet to help us assess the overall quality before approval so ensure it has been proofread, spellchecked and grammarchecked.

Question-Answering Web Apps have been around for a while now. A good example is the use of bots on websites where a user asks a question and an answer is automatically generated. In this build, I will show our readers how to build a Q&A app using one of the pre-trained models on HuggngFace. Anvil is an open-source web-interface that allows a developer to spin up and deploy a web app using Python. It is quite similar to the popular [Gradio](https://gradio.app/) web application used by developers. We will leverage Anvil to build our front-end.

#### Key takeaways 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
Use this area to get your ideas down on the bulk of your article or tutorial.

By the end of this tutorial, a reader will be able to understand:
1. Where to fetch a pre-trained BERT model from.
2. The different BERT models available.
3. How to import and train the BERT model for Q&A.
4. How to leverage [Anvil](https://anvil.works/) to make the app user friendly.

#### Article quality
How is your article unique? Tell us what makes your approach different from similar articles that have published on the same topic?
Is yours more in-depth? Does it cover additional topics? Do you provide handy tips or anecdotal advice?

There have been articles before that have build Q&A web apps using Python mostly using the BERT model. However, there are many BERT models with different hyper-parameters that exist today. I will be using the multilingual BERT model for this build. This isn't a commonly used model. In addition, most articles use Gradio as the web interface of choice. I will be using [Anvil](https://anvil.works/) instead.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.
- [Anvil](https://anvil.works/)

### PR DESCRIPTION
**Pull request by lcheptoo**
This closes #6263

## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by LinusMuema** [comment]
Hi @lcheptoo 

Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[How to build a Q&A web application using Python and Anvil #6394 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7989196/How.to.build.a.Q.A.web.application.using.Python.and.Anvil.6394.Copyleaks.Report.pdf)

**Comment by lcheptoo** [comment]
> Hi @lcheptoo
> 
> Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.
> 
> [How to build a Q&A web application using Python and Anvil #6394 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7989196/How.to.build.a.Q.A.web.application.using.Python.and.Anvil.6394.Copyleaks.Report.pdf)

@lcheptoo I know it looks like plagiarized text but it's actually part of the tutorial. Well. I was trying to demonstrate how using a passage fetched from Wikipedia, you can feed that passage into the model and it could actually answer some of the questions you ask it. For example, "What is machine learning?" The model can go through the passage and fetch answers.

File: content/articles/building-a-qa-web-application/index.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
+Question-Answering Web Apps have been around for a while now. A good example is the use of bots on websites where a user asks a question and an answer is automatically generated. In this build, we will show our readers how to build a Q&A app using one of the pre-trained models on [HuggingFace](https://huggingface.co/).
```
Comment (Bayler) [inline content/articles/building-a-qa-web-application/index.md:17]:
```suggestion
Question-Answering Web Apps have been around for a while now. A good example is the use of bots on websites where a user asks a question, and an answer is generated automatically. In this build, we will show our readers how to build a Q&A app using one of the pre-trained models on [HuggingFace](https://huggingface.co/).
```

File: content/articles/building-a-qa-web-application/index.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+```bash
+!pip install transformers
+```
+Those are the two main dependencies that we will use in this tutorial. Let's go ahead and import the tokenizer, pipeline, and model from transformers.
+
+```python
+from transformers import AutoModelForQuestionAnswering, AutoTokenizer, pipeline
+```
+- The `AutoTokenizer` allows us to convert natural language into tokens which can then be used by the NLP model.
```
Comment (Bayler) [inline content/articles/building-a-qa-web-application/index.md:53]:
```suggestion
- The NLP model uses the `AutoTokenizer` to convert natural language into tokens.
```

File: content/articles/building-a-qa-web-application/index.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+}
+```
+Results:
+```bash
+{'answer': 'medicine, email filtering, speech recognition, and computer vision',
+ 'end': 509,
+ 'score': 0.27500787377357483,
+ 'start': 443}
+```
+The model seems to be giving us the right answers by simply fetching context from the paragraph. This is so nice! We could take this a step further a build a web application that enables users to interact with the model easily. This is where Anvil comes in handy.
```
Comment (Bayler) [inline content/articles/building-a-qa-web-application/index.md:111]:
```suggestion
The model seems to be giving us the answers by simply fetching context from the paragraph. This is so nice! We can take this a step further. We can build a web interface that enables users to interact with the model easily. This is where Anvil comes in handy.
```

**Comment by Bayler** [comment]
@lcheptoo Very educative! Please work on the few revisions before we can move on to final review.

**Comment by lcheptoo** [comment]
@Bayler done making the changes.

**Comment by LinusMuema** [comment]
jumping on this one for the final review

**Comment by LinusMuema** [comment]
Hi @lcheptoo 

Great job on the article

**Pull request merged**